src/System.ServiceModel.Federation/src/System/ServiceModel/Federation/IWSTrustContract.cs

```diff
@@ -0,0 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ServiceModel.Channels;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Federation
+{
+    /// <summary>
+    /// Defines the IWSTrustContract for sending WS-Trust messages to an STS.
+    /// </summary>
+    [ServiceContract]
+    public interface IWSTrustContract
+    {
+        /// <summary>
+        /// Method for Cancel binding for WS-Trust.
+        /// </summary>
+        /// <param name="message">The WS-Trust request to send to an STS.</param>
+        /// <returns>A message containing a WS-Trust response.</returns>
+        [OperationContract(Name = "Cancel", Action = "*", ReplyAction = "*")]
+        Task<Message> CancelAsync(Message message);
+
```

Comment:
All of these methods are interchangeable as they take and return a Message and they all have the Action and ReplyAction to be "*". I need to spend a little bit of time trying some things with this code, but I'm pretty certain that the Name property is only used to provide the Action and ReplyAction if they aren't explicitly specified. This whole interface seems like it could be eliminated and replaced with a Task based equivalent of IRequestChannel. IWSTrustChannelContract has the strongly typed operations on it. I'll update once I know more.

Reply:
I have done some local testing and confirmed what I believed. When an operation is defined like this, the is zero request shape that comes from the operation. Every method in this interface will generate the exact same on the wire message. I have no idea what the original intention was of the authors of the class in NetFx, but this interface make no sense at all to exist. We can remove this interface and just have IWSTrustChannelContract live on it's own.

Reply:
It does make a difference, the responsibility is on the user to create the correct message.
To make it easier for users we probably would want to make WSTrustBodyWriter public, although the code is simple.

see code here: https://github.com/brentschmaltz/WCF/blob/9b084e846fea828fbb5d65bbe8ac70bdd26f9ded/WsTrustClientCore/Program.cs#L46

SelfHostSTS will identify the different messages and route them.
For Renew see: https://github.com/brentschmaltz/WCF/blob/9b084e846fea828fbb5d65bbe8ac70bdd26f9ded/SelfHostSTS/SelfHostSecurityTokenService.cs#L215